python/pyspark/sql/dataframe.py

```diff
@@ -3008,6 +3008,34 @@ def __getattr__(self, name: str) -> Column:
         jc = self._jdf.apply(name)
         return Column(jc)
 
+    def __dir__(self) -> List[str]:
+        """
+        Examples
+        --------
+        >>> from pyspark.sql.functions import lit
+
+        Create a dataframe with a column named 'id'.
+
+        >>> df = spark.range(3)
+        >>> [attr for attr in dir(df) if attr[0] == 'i'][:7] # Includes column id
+        ['id', 'inputFiles', 'intersect', 'intersectAll', 'isEmpty', 'isLocal', 'isStreaming']
+
+        Add a column named 'i_like_pancakes'.
+
+        >>> df = df.withColumn('i_like_pancakes', lit(1))
+        >>> [attr for attr in dir(df) if attr[0] == 'i'][:7] # Includes columns i_like_pancakes, id
+        ['i_like_pancakes', 'id', 'inputFiles', 'intersect', 'intersectAll', 'isEmpty', 'isLocal']
+
+        Try to add an existed column 'inputFiles'.
+
+        >>> df = df.withColumn('inputFiles', lit(2))
+        >>> [attr for attr in dir(df) if attr[0] == 'i'][:7] # Doesn't duplicate inputFiles
+        ['i_like_pancakes', 'id', 'inputFiles', 'intersect', 'intersectAll', 'isEmpty', 'isLocal']
+        """
+        attrs = list(super().__dir__())
+        attrs.extend(attr for attr in self.columns if attr not in attrs)
```

Comment:
can we use `hasattr` instead? That will use `getattr`, and invoke `__getattr__` in the dataframe.

Reply:
I don't see the point of using hasattr: our implementation about `__getattr__ `check the columns exist or not and we still  need to check that `attr not in attrs`

Reply:
I wanted to make sure that we share the same logic when we get the attributes. e.g., this would matter in the maintenance and it's future proof - we won't have to deduplicately fix both places. But this is pretty simple logic so I am fine as is.